File: src/gimpish/geometry.py
"""Pure geometry/color helpers — no pyvips dependency, easy to unit-test.

Semantic placement verbs (fit/fill/cover + anchor) resolve to concrete pixel
transforms here; the agent expresses intent and we write pixels into the scene.
"""

from __future__ import annotations

from typing import Tuple

RGBA = Tuple[int, int, int, int]

# anchor -> (fx, fy) where 0 = left/top, 0.5 = center, 1 = right/bottom
ANCHORS: dict[str, Tuple[float, float]] = {
    "top-left": (0.0, 0.0),
    "top": (0.5, 0.0),
    "top-right": (1.0, 0.0),
    "left": (0.0, 0.5),
    "center": (0.5, 0.5),
    "right": (1.0, 0.5),
    "bottom-left": (0.0, 1.0),
    "bottom": (0.5, 1.0),
    "bottom-right": (1.0, 1.0),
}


def parse_color(text: str) -> RGBA:
    """Parse '#rgb', '#rrggbb', '#rrggbbaa' (alpha optional, defaults opaque)."""
    s = text.strip().lstrip("#")
    if len(s) == 3:
        s = "".join(ch * 2 for ch in s)
    if len(s) == 6:
        s += "ff"
    if len(s) != 8:
        raise ValueError(f"bad color {text!r}: expected #rgb, #rrggbb, or #rrggbbaa")
    try:
        r, g, b, a = (int(s[i : i + 2], 16) for i in (0, 2, 4, 6))
    except ValueError as exc:
        raise ValueError(f"bad color {text!r}") from exc
    return (r, g, b, a)


def resolve_fit(
    src_w: int,
    src_h: int,
    canvas_w: int,
    canvas_h: int,
    mode: str,
    percent: float = 100.0,
    anchor: str = "center",
) -> Tuple[float, float, float]:
    """Return (scale, x, y) placing a src_w x src_h image on the canvas.

    mode:
      fit   -> contain inside a box of (percent% of canvas), centered/anchored
      fill  -> cover the whole canvas (percent scales the cover), overflow cropped
      cover -> alias of fill
    """
    frac = percent / 100.0
    if mode == "fit":
        box_w = canvas_w * frac
        box_h = canvas_h * frac
        scale = min(box_w / src_w, box_h / src_h)
    elif mode in ("fill", "cover"):
        scale = max(canvas_w / src_w, canvas_h / src_h) * frac
    else:
        raise ValueError(f"unknown fit mode {mode!r} (use fit|fill|cover)")

    scaled_w = src_w * scale
    scaled_h = src_h * scale
    fx, fy = ANCHORS.get(anchor, (0.5, 0.5))
    x = (canvas_w - scaled_w) * fx
    y = (canvas_h - scaled_h) * fy
    return scale, x, y


def parse_stops(text: str) -> list[dict]:
    """Parse 'at:color, at:color, ...' e.g. '0:#000000ff, 1:#00000000'."""
    stops = []
    for part in text.split(","):
        part = part.strip()
        if not part:
            continue
        at_str, _, color = part.partition(":")
        if not color:
            raise ValueError(f"bad gradient stop {part!r}: expected 'position:#color'")
        stops.append({"at": float(at_str), "color": color.strip()})
    if len(stops) < 2:
        raise ValueError("gradient needs at least 2 stops")
    return sorted(stops, key=lambda s: s["at"])

File: packages/core/src/geometry.ts
/** Canvas-space geometry: anchors, fit resolution, gradient stops (see DESIGN.md §5). */

export const ANCHORS = {
  "top-left": [0, 0],
  top: [0.5, 0],
  "top-right": [1, 0],
  left: [0, 0.5],
  center: [0.5, 0.5],
  right: [1, 0.5],
  "bottom-left": [0, 1],
  bottom: [0.5, 1],
  "bottom-right": [1, 1],
} as const satisfies Record<string, readonly [number, number]>;

export type Anchor = keyof typeof ANCHORS;

export function anchorFractions(anchor: string): readonly [number, number] {
  return (ANCHORS as Record<string, readonly [number, number]>)[anchor] ?? [0.5, 0.5];
}

export type FitMode = "fit" | "fill" | "cover";

export interface FitResult {
  scale: number;
  x: number;
  y: number;
}

/**
 * Place a srcW x srcH image on the canvas.
 *
 * fit   -> contain inside a box of (percent% of canvas), anchored
 * fill  -> cover the whole canvas (percent scales the cover), overflow cropped
 * cover -> alias of fill
 */
export function resolveFit(
  srcW: number,
  srcH: number,
  canvasW: number,
  canvasH: number,
  mode: FitMode,
  percent = 100,
  anchor = "center",
): FitResult {
  const frac = percent / 100;
  let scale: number;
  if (mode === "fit") {
    scale = Math.min((canvasW * frac) / srcW, (canvasH * frac) / srcH);
  } else if (mode === "fill" || mode === "cover") {
    scale = Math.max(canvasW / srcW, canvasH / srcH) * frac;
  } else {
    throw new Error(`unknown fit mode ${JSON.stringify(mode)} (use fit|fill|cover)`);
  }
  const [fx, fy] = anchorFractions(anchor);
  return {
    scale,
    x: (canvasW - srcW * scale) * fx,
    y: (canvasH - srcH * scale) * fy,
  };
}

export interface GradientStop {
  at: number;
  color: string;
}

/** Parse 'at:color, at:color, ...' e.g. '0:#000000ff, 1:#00000000'. */
export function parseStops(text: string): GradientStop[] {
  const stops: GradientStop[] = [];
  for (const rawPart of text.split(",")) {
    const part = rawPart.trim();
    if (!part) continue;
    const sep = part.indexOf(":");
    if (sep < 0) {
      throw new Error(`bad gradient stop ${JSON.stringify(part)}: expected 'position:#color'`);
    }
    const at = Number.parseFloat(part.slice(0, sep));
    const color = part.slice(sep + 1).trim();
    if (!Number.isFinite(at) || !color) {
      throw new Error(`bad gradient stop ${JSON.stringify(part)}: expected 'position:#color'`);
    }
    stops.push({ at, color });
  }
  if (stops.length < 2) throw new Error("gradient needs at least 2 stops");
  return stops.sort((a, b) => a.at - b.at);
}

/** Rotate (x, y) about (px, py) by `degrees` clockwise (screen-space y-down). */
export function rotatePoint(
  x: number,
  y: number,
  px: number,
  py: number,
  degrees: number,
): { x: number; y: number } {
  const r = (degrees * Math.PI) / 180;
  const c = Math.cos(r);
  const s = Math.sin(r);
  const dx = x - px;
  const dy = y - py;
  return { x: px + dx * c - dy * s, y: py + dx * s + dy * c };
}
